File: leetcode/p00232/p00232.go
package p00232

type MyQueue struct {
	input  Stack
	output Stack
}

func Constructor() MyQueue {
	return MyQueue{
		input:  CreateStack(),
		output: CreateStack(),
	}
}

func (this *MyQueue) Push(x int) {
	this.input.Push(x)
}

func (this *MyQueue) Pop() int {
	if this.Empty() {
		return -1
	}

	if this.output.IsEmpty() {
		for !this.input.IsEmpty() {
			this.output.Push(this.input.Pop())
		}
	}

	return this.output.Pop()
}

func (this *MyQueue) Peek() int {
	if this.Empty() {
		return -1
	}

	if this.output.IsEmpty() {
		for !this.input.IsEmpty() {
			this.output.Push(this.input.Pop())
		}
	}

	return this.output.Peek()
}

func (this *MyQueue) Empty() bool {
	return this.input.Size()+this.output.Size() == 0
}

type Stack struct {
	arr  []int
	size int
}

func CreateStack() Stack {
	return Stack{
		arr:  make([]int, 0),
		size: 0,
	}
}
func (s *Stack) IsEmpty() bool {
	return s.size == 0
}

func (s *Stack) Size() int {
	return s.size
}

func (s *Stack) Push(val int) {
	s.arr = append(s.arr, val)
	s.size++
}

func (s *Stack) Peek() int {
	if s.IsEmpty() {
		return -1
	}

	return s.arr[s.size-1]
}

func (s *Stack) Pop() int {
	if s.IsEmpty() {
		return -1
	}

	val := s.arr[s.size-1]
	s.arr = s.arr[:s.size-1]
	s.size--

	return val
}

File: leetcode/p00232/p00232.py
from collections import deque

class MyQueue:

    def __init__(self):
        self.input = deque()
        self.output = deque()

    def push(self, x: int) -> None:
        self.input.append(x)

    def pop(self) -> int:
        if self.empty():
            return -1
        
        if len(self.output) == 0:
            while self.input:
                self.output.append(self.input.pop())
        
        return self.output.pop()

    def peek(self) -> int:
        if self.empty():
            return -1
        
        if len(self.output) == 0:
            while self.input:
                self.output.append(self.input.pop())
            
        return self.output[-1]


    def empty(self) -> bool:
        return True if len(self.input) + len(self.output) == 0 else False

if __name__ == '__main__':
    obj = MyQueue()
    obj.push(1)
    obj.push(2)
    assert obj.peek() == 1
    assert obj.pop() == 1
    assert obj.empty() == False
